clang/docs/checkedc/Bounds-Widening-for-Null-Terminated-Arrays.md

```diff
@@ -2,165 +2,290 @@
-null-terminated arrays.
+The bounds of a null-terminated array can also be widened when the initial
+elements of the sequence are implicitly read (for example via call to `strlen`
+function). In the example below the bounds of the null-terminated-array `p` are
+widened to `bounds(p, p + x)`.
 
-## Dataflow Analysis Properties
-The key properties of the dataflow analysis are described below:
+```
+  void f(_Nt_array_ptr<char> p) {
+    int x = strlen(p) _Where p : bounds(p, p + x);
+  }
+```
 
-1. **Forward:** The basic blocks of the function are traversed in reverse
-post-order. In other words, a basic block is visited before its successors.
+In the next section we describe a dataflow analysis to widen bounds for
+null-terminated arrays. The dataflow analysis is **forward**,
+**path-sensitive**, **flow-sensitive** and **intra-procedural**.
 
-2. **Path-sensitive:** Path-sensitivity means that the dataflow analysis
-generates different facts on the `then` and `else` branches.
+Note: This analysis uses the bounds declarations in where clauses as facts and
+assumes that they have already been proven.
 
-3. **Flow-sensitive:** Flow-sensitivity means that the sequence of instructions
-in a basic block is taken into consideration when performing the analysis.
+## Dataflow Analysis for Widening the Bounds of Null-terminated Arrays
+We use `V` to denote a null-terminated array variable, `bounds(Lower, Upper)`
+to denote the bounds expression for `V`, `S` to denote a statement, and `B` and
+`B'` to denote basic blocks.
 
-4. **Intra-procedural:** The analysis is done on one function at a time.
+Note: The analysis treats two variables having the same name but different
+declarations as distinct from each other.
 
-## Dataflow Analysis Details
-For every basic block, we compute the following sets: `In` and `Kill`. The `In`
-set for basic block `B` is denoted as `In[B]` and the `Kill` set is denoted as
-`Kill[B]`.
+The dataflow analysis tracks all null-terminated array variables local to a
+function along with their bounds expressions. The dataflow facts that flow
+through the analysis are sets of pairs `V:bounds(Lower, Upper)`.
 
-For every edge, we compute the following sets: `Out` and `Gen`. The `Out` set on
-edge `Bi->Bj` is denoted as `Out[Bi][Bj]` and the `Gen` set is denoted as
-`Gen[Bi][Bj]`.
+For every basic block `B`, we compute the sets `In[B]` and `Out[B]`.
 
-### In[B]
-`In[B]` stores the mapping between an `_Nt_array_ptr` and its widened bounds
-inside block `B`. For example, given `_Nt_array_ptr V` with declared bounds
-`(V + low, V + high)`, `In[B]` would store the mapping `{V:i}`, where `i` is an
-unsigned integer implying that the bounds of `V` should be widened to
-`(V + low, V + high + i)`.
+For every statement `S`, we compute the sets `Gen[S]` and `Kill[S]`.
+
+The sets `In[B]` and `Out[B]` are part of the fixed-point computation, whereas
+the sets `Gen[S]` and `Kill[S]` are computed **before** the fixed-point
```

Comment:
I believe Gen[S] and Kill[S] should be properties of basic blocks, not statements.   This follows the convention for other dataflow analyses, where you compute summary information for a basic block.